stdlib/public/core/NFD.swift

```diff
@@ -0,0 +1,220 @@
+    let hashedScalar = (decompEntry << 14) >> 14
+
+    // If this is not our original scalar, then we have no decomposition for this
+    // scalar, so just emit itself.
+    guard scalar.value == hashedScalar else {
+      buffer.append((scalar: scalar, normData: normData))
+      return
+    }
+    
+    // Otherwise, our index is stored in the top 14 bits.
+    let decompIdx = Int(decompEntry >> 18)
+    
+    let decompPtr = _swift_stdlib_nfd_decompositions._unsafelyUnwrappedUnchecked
+    
+    // This size is the utf8 length of the decomposition.
+    var size = Int(decompPtr[decompIdx])
+    
+    let utf8 = UnsafeBufferPointer(
+      start: decompPtr + decompIdx + 1,
+      count: size
+    )
+
+    while size > 0 {
+      let (scalar, len) = _decodeScalar(utf8, startingAt: utf8.count &- size)
+      size &-= len
+
+      let normData: UInt16
+
+      // Fast path: Because this will be emitted into the completed NFD buffer,
+      // we don't need to look at NFD_QC anymore which lets us do a larger
+      // latiny check for NFC_QC and CCC (0xC0 vs. 0x300).
+      if scalar.value < 0x300 {
+        normData = 0
+      } else {
+        normData = _swift_stdlib_getNormData(scalar.value)
+      }
+
+      buffer.append((scalar: scalar, normData: normData))
+    }
+  }
+
+  internal mutating func next() -> (scalar: Unicode.Scalar, normData: UInt16)? {
+    // Empty out our buffer before attempting to decompose the next
+    // normalization segment.
+    if !buffer.isEmpty {
+      return buffer.removeLast()
+    }
+    
+    // If we have a starter stored, take care of that before asking for more
+    // scalars.
+    if let starter = starter.take() {
+      let ccc = starter.normData >> 3
+      var isNFDQC = starter.normData & 1 == 0
+      
+      // Because we don't store precomposed hangul in our NFD_QC data, these
+      // will return true for NFD_QC when in fact they are not.
+      if (0xAC00 ... 0xD7A3).contains(starter.scalar.value) {
+        isNFDQC = false
+      }
+      
+      // If our scalar IS NFD quick check, then it's as simple as appending to
+      // our buffer and moving on the next scalar. Otherwise, we need to
+      // decompose this and append each decomposed scalar.
+      if isNFDQC {
+        // Fast path: If our scalar is also ccc = 0, then this doesn't need to
+        // be appended to the buffer at all.
+        if ccc == 0 {
+          return starter
+        }
+
+        buffer.append(starter)
+      } else {
+        decompose(starter.scalar, with: starter.normData)
+      }
+    }
+    
+    while let scalar = iterator.next() {
+      let normData = _swift_stdlib_getNormData(scalar.value)
+      let ccc = normData >> 3
+      var isNFDQC = normData & 1 == 0
+      
+      // Because we don't store precomposed hangul in our NFD_QC data, these
+      // will return true for NFD_QC when in fact they are not.
+      if (0xAC00 ... 0xD7A3).contains(scalar.value) {
+        isNFDQC = false
+      }
```

Comment:
It would be really nice to refactor this into the computed property on `_NormData`.